File: src/main/java/discretemath/structure/set/multi/MultiSet.java
package discretemath.structure.set.multi;

import java.util.Iterator;

import discretemath.structure.set.Set;
import discretemath.structure.tuple.Pair;

public interface MultiSet<T> extends Set<T> {

	MultiSet<T> sumWith(MultiSet<T> set);

	T mode();

	int itemCount(T item);

	Iterator<Pair<T, Integer>> withCountIterator();
}

File: src/main/scala/discretemath/structure/set/multi/MultiSet.scala
package discretemath.structure.set.multi

import discretemath.structure.set.BaseSet

trait MultiSet[T] extends BaseSet[T, MultiSet, MultiSet[T]] {

  def sumWith(set: MultiSet[T]): MultiSet[T]

  def mode: Option[T]

  def itemCount(item: T): Int

  def withCountIterator: Iterator[(T, Int)]
}